src/main/java/redis/clients/jedis/JedisClusterConnectionHandler.java

```diff
@@ -34,12 +34,22 @@ public Jedis getConnectionFromNode(HostAndPort node) {
 
   private void initializeSlotsCache(Set<HostAndPort> startNodes, GenericObjectPoolConfig poolConfig) {
     for (HostAndPort hostAndPort : startNodes) {
-      Jedis jedis = new Jedis(hostAndPort.getHost(), hostAndPort.getPort());
+      Jedis jedis = null;
       try {
+        jedis = new Jedis(hostAndPort.getHost(), hostAndPort.getPort());
         cache.discoverClusterNodesAndSlots(jedis);
         break;
       } catch (JedisConnectionException e) {
-        // try next nodes
+        // close already created resource and try next nodes
+        for (JedisPool pool : cache.getNodes().values()) {
```

Comment:
Imagine situation. We got first node, we discover something and fails in the middle. Now we "try next node" and just make .clear() .clear() on both collection in cache because we will discover everything from this next node. Without releasing allocated resources, connections, etc. So, it's just this releasing.